#### Demographics and Employment in the US
The government measures unemployment using the Current Population Survey (CPS). In this exercise, we will use a compact form of the September 2013 version of this rich, nationally representative dataset (usually available online).

In [1]:
##set up
import pandas as pd
import numpy as np
import matplotlib as plt
%matplotlib inline

In [2]:
#Read in data:
data=pd.read_csv('CPSData.csv')
data.head()

PeopleInHousehold Region    State  MetroAreaCode  Age        Married  \
0                  1  South  Alabama        26620.0   85        Widowed   
1                  3  South  Alabama        13820.0   21  Never Married   
2                  3  South  Alabama        13820.0   37  Never Married   
3                  3  South  Alabama        13820.0   18  Never Married   
4                  3  South  Alabama        26620.0   52        Widowed   

      Sex               Education   Race  Hispanic  CountryOfBirthCode  \
0  Female        Associate degree  White         0                  57   
1    Male             High school  Black         0                  57   
2  Female             High school  Black         0                  57   
3    Male  No high school diploma  Black         0                  57   
4  Female        Associate degree  White         0                  57   

       Citizenship    EmploymentStatus                            Industry  
0  Citizen, Native             Retired                                 NaN  
1  Citizen, Native          Unemployed  Professional and business services  
2  Citizen, Native            Disabled                                 NaN  
3  Citizen, Native  Not in Labor Force                                 NaN  
4  Citizen, Native            Employed  Professional and business services

#### How many interviewees int his dataset?
Among the interviewees with a value reported for the Industry variable, what is the most common industry of employment?

In [3]:
data.shape

(131302, 14)

In [4]:
data.Industry.value_counts()

Educational and health services                15017
Trade                                           8933
Professional and business services              7519
Manufacturing                                   6791
Leisure and hospitality                         6364
Construction                                    4387
Financial                                       4347
Transportation and utilities                    3260
Other services                                  3224
Public administration                           3186
Information                                     1328
Agriculture, forestry, fishing, and hunting     1307
Mining                                           550
Armed forces                                      29
Name: Industry, dtype: int64

#### * Which state has the fewest interviewees?
#### * Which state has the largest number of interviewees?
#### * What proportion of interviewees are citizens of the United States?



In [5]:
(data.groupby('State').count())['Region'].sort_values()

State
New Mexico               1102
Montana                  1214
Mississippi              1230
Alabama                  1376
West Virginia            1409
Arkansas                 1421
Louisiana                1450
Idaho                    1518
Oklahoma                 1523
Arizona                  1528
Alaska                   1590
Wyoming                  1624
North Dakota             1645
South Carolina           1658
Tennessee                1784
District of Columbia     1791
Kentucky                 1841
Utah                     1842
Nevada                   1856
Vermont                  1890
Kansas                   1935
Oregon                   1943
Nebraska                 1949
Massachusetts            1987
South Dakota             2000
Indiana                  2004
Hawaii                   2099
Missouri                 2145
Rhode Island             2209
Delaware                 2214
Maine                    2263
Washington               2366
Iowa                     2528
New 

* Least- New Mexico
* Largest num of Interviewees: California

In [6]:
(data['Citizenship'].value_counts())
float(116639+7073)/131302 

0.9421943306271039

Citizens are 94.2% of the data.

#### The CPS differentiates between race (with possible values American Indian, Asian, Black, Pacific Islander, White, or Multiracial) and ethnicity. A number of interviewees are of Hispanic ethnicity, as captured by the Hispanic variable. For which races are there at least 250 interviewees in the CPS dataset of Hispanic ethnicity? 

In [7]:
data.groupby('Race')['Hispanic'].sum()
#data['Hispanic']

Race
American Indian       304
Asian                 113
Black                 621
Multiracial           448
Pacific Islander       77
White               16731
Name: Hispanic, dtype: int64

#### Which variables have at least one interviewee with a missing (NA) value? 

In [8]:
data.isnull().sum()

PeopleInHousehold         0
Region                    0
State                     0
MetroAreaCode         34238
Age                       0
Married               25338
Sex                       0
Education             25338
Race                      0
Hispanic                  0
CountryOfBirthCode        0
Citizenship               0
EmploymentStatus      25789
Industry              65060
dtype: int64

#### Evaluating Missing Values:
Lets try to determine if there is a pattern in the missing values of the Married variable. Lets try grouping it by different variables

In [9]:
data.groupby('Region')['Married'].value_counts()

Region     Married      
Midwest    Married          13272
           Never Married     6869
           Divorced          2630
           Widowed           1505
           Separated          333
Northeast  Married          11078
           Never Married     6508
           Divorced          2122
           Widowed           1332
           Separated          392
South      Married          17301
           Never Married     9606
           Divorced          3586
           Widowed           2221
           Separated          821
West       Married          13858
           Never Married     7789
           Divorced          2813
           Widowed           1447
           Separated          481
Name: Married, dtype: int64

In [10]:
###Aah...but what I need is to group ONLY the null values of Married by region..
##NOTE: I cannot use the column married  as he its values are NAN and so the count..gives me 0
##So any other field will work..
(data[data.Married.isnull()]).groupby('Region')['Sex'].count()

Region
Midwest      6075
Northeast    4507
South        7967
West         6789
Name: Sex, dtype: int64

In [11]:
(data[data.Married.isnull()]).groupby('Region')['PeopleInHousehold'].count()

Region
Midwest      6075
Northeast    4507
South        7967
West         6789
Name: PeopleInHousehold, dtype: int64

In [12]:
data[data['Married'].isnull()].groupby('Sex')['Region'].count()

Sex
Female    12217
Male      13121
Name: Region, dtype: int64

In [13]:
data[data.Married.isnull()].groupby('Age')['Region'].count()

Age
0     1283
1     1559
2     1574
3     1693
4     1695
5     1795
6     1721
7     1681
8     1729
9     1748
10    1750
11    1721
12    1797
13    1802
14    1790
Name: Region, dtype: int64

In [14]:
data[data.Married.isnull()]

PeopleInHousehold Region    State  MetroAreaCode  Age Married     Sex  \
13                      4  South  Alabama        26620.0    2     NaN  Female   
14                      4  South  Alabama        26620.0    4     NaN    Male   
17                      2  South  Alabama        13820.0   13     NaN  Female   
27                      3  South  Alabama        33860.0    2     NaN  Female   
34                      6  South  Alabama        33860.0    3     NaN  Female   
35                      6  South  Alabama        33860.0   11     NaN  Female   
47                      2  South  Alabama        26620.0    5     NaN    Male   
50                      4  South  Alabama        33660.0   14     NaN    Male   
56                      3  South  Alabama        13820.0    5     NaN    Male   
57                      3  South  Alabama        13820.0   11     NaN  Female   
59                      5  South  Alabama        33860.0    9     NaN    Male   
60                      5  South  Alabama        33860.0   12     NaN  Female   
61                      5  South  Alabama        33860.0   14     NaN    Male   
84                      5  South  Alabama        13820.0   10     NaN    Male   
85                      5  South  Alabama        13820.0   10     NaN    Male   
91                      3  South  Alabama        26620.0    9     NaN    Male   
99                      2  South  Alabama        26620.0    1     NaN  Female   
123                     3  South  Alabama        33860.0    3     NaN    Male   
136                     4  South  Alabama        33660.0   14     NaN  Female   
149                     3  South  Alabama        33660.0    2     NaN    Male   
169                     2  South  Alabama        13820.0    1     NaN  Female   
188                     5  South  Alabama        33660.0    1     NaN    Male   
189                     5  South  Alabama        33660.0    4     NaN  Female   
190                     5  South  Alabama        33660.0    7     NaN  Female   
193                     3  South  Alabama        13820.0    6     NaN  Female   
202                     4  South  Alabama        26620.0    3     NaN    Male   
203                     4  South  Alabama        26620.0    5     NaN    Male   
204                     4  South  Alabama        26620.0    8     NaN  Female   
212                     2  South  Alabama        33660.0    9     NaN  Female   
228                     3  South  Alabama        13820.0    6     NaN  Female   
...                   ...    ...      ...            ...  ...     ...     ...   
131174                  5   West  Wyoming            NaN   11     NaN    Male   
131177                  4   West  Wyoming            NaN    7     NaN    Male   
131182                  5   West  Wyoming            NaN    8     NaN  Female   
131183                  5   West  Wyoming            NaN   11     NaN    Male   
131184                  5   West  Wyoming            NaN   14     NaN  Female   
131189                  4   West  Wyoming            NaN   10     NaN  Female   
131190                  4   West  Wyoming            NaN   12     NaN  Female   
131197                  3   West  Wyoming            NaN    2     NaN    Male   
131200                  4   West  Wyoming            NaN   14     NaN    Male   
131205                  4   West  Wyoming            NaN    4     NaN  Female   
131206                  4   West  Wyoming            NaN    7     NaN    Male   
131209                  4   West  Wyoming            NaN    1     NaN    Male   
131210                  4   West  Wyoming            NaN    5     NaN    Male   
131211                  4   West  Wyoming            NaN   14     NaN    Male   
131222                  3   West  Wyoming            NaN   12     NaN    Male   
131232                  3   West  Wyoming            NaN    9     NaN    Male   
131233                  3   West  Wyoming            NaN   12     NaN  Female   
131240                  6   West  Wyomi

In [15]:
data[data.Married.isnull()].groupby('Citizenship')['Region'].count()

Citizenship
Citizen, Native         24683
Citizen, Naturalized      163
Non-Citizen               492
Name: Region, dtype: int64

#### So Marriage variable is missing for all interviewees between the age 0-14

#### MetroAreaCode is missing if an interviewee does not live in a metropolitan area. 
* How many states had all interviewees living in a non-metropolitan area (aka they have a missing MetroAreaCode value)? For this question, treat the District of Columbia as a state (even though it is not technically a state).
* How many states had all interviewees living in a metropolitan area? Again, treat the District of Columbia as a state.
* 

In [16]:
data[data.MetroAreaCode.isnull()].groupby('State').count()

PeopleInHousehold  Region  MetroAreaCode   Age  Married   Sex  \
State                                                                           
Alabama                       356     356              0   356      300   356   
Alaska                       1590    1590              0  1590     1246  1590   
Arizona                       201     201              0   201      162   201   
Arkansas                      697     697              0   697      548   697   
California                    237     237              0   237      185   237   
Colorado                      380     380              0   380      305   380   
Connecticut                   243     243              0   243      203   243   
Delaware                      518     518              0   518      449   518   
Florida                       202     202              0   202      172   202   
Georgia                       557     557              0   557      433   557   
Hawaii                        523     523              0   523      400   523   
Idaho                         757     757              0   757      586   757   
Illinois                      439     439              0   439      345   439   
Indiana                       584     584              0   584      484   584   
Iowa                         1231    1231              0  1231     1026  1231   
Kansas                        701     701              0   701      574   701   
Kentucky                      933     933              0   933      782   933   
Louisiana                     234     234              0   234      185   234   
Maine                        1354    1354              0  1354     1119  1354   
Maryland                      222     222              0   222      177   222   
Massachusetts                 129     129              0   129      110   129   
Michigan                      546     546              0   546      451   546   
Minnesota                     989     989              0   989      788   989   
Mississippi                   854     854              0   854      696   854   
Missouri                      705     705              0   705      561   705   
Montana                      1015    1015              0  1015      852  1015   
Nebraska                     1133    1133              0  1133      925  1133   
Nevada                        247     247              0   247      202   247   
New Hampshire                1514    1514              0  1514     1275  1514   
New Mexico                    270     270              0   270      207   270   
New York                      451     451              0   451      363   451   
North Carolina                977     977              0   977      805   977   
North Dakota                 1213    1213              0  1213      992  1213   
Ohio                          924     924              0   924      752   924   
Oklahoma                      499     499              0   499      386   499   
Oregon                        424     424              0   424      347   424   
Pennsylvania                  685     685              0   685      580   685   
South Carolina                519     519              0   519      445   519   
South Dakota                 1405    1405              0  1405     1131  1405   
Tennessee                     635     635              0   635      529   635   
Texas                        1017    1017              0  1017      774  1017   
Utah                          387     387              0   387      271   387   
Vermont                      1233    1233              0  1233     1045  1233   
Virginia                      586     586              0   586      485   586   
Washington                    429     429              0   429      358   429   
West Virginia                1065    1065              0  1065      898  1065   
Wisconsin                     804     804              0   804      641   804   
Wyoming                      1624    1624      

In [17]:
data[data.MetroAreaCode.isnull()].groupby('State')['State'].count()

State
Alabama            356
Alaska            1590
Arizona            201
Arkansas           697
California         237
Colorado           380
Connecticut        243
Delaware           518
Florida            202
Georgia            557
Hawaii             523
Idaho              757
Illinois           439
Indiana            584
Iowa              1231
Kansas             701
Kentucky           933
Louisiana          234
Maine             1354
Maryland           222
Massachusetts      129
Michigan           546
Minnesota          989
Mississippi        854
Missouri           705
Montana           1015
Nebraska          1133
Nevada             247
New Hampshire     1514
New Mexico         270
New York           451
North Carolina     977
North Dakota      1213
Ohio               924
Oklahoma           499
Oregon             424
Pennsylvania       685
South Carolina     519
South Dakota      1405
Tennessee          635
Texas             1017
Utah               387
Vermont           1233
Virgi

In [18]:
metroisnull=list(data[data.MetroAreaCode.isnull()].groupby('State')['State'].count().index)

In [19]:
metroisnull

['Alabama',
 'Alaska',
 'Arizona',
 'Arkansas',
 'California',
 'Colorado',
 'Connecticut',
 'Delaware',
 'Florida',
 'Georgia',
 'Hawaii',
 'Idaho',
 'Illinois',
 'Indiana',
 'Iowa',
 'Kansas',
 'Kentucky',
 'Louisiana',
 'Maine',
 'Maryland',
 'Massachusetts',
 'Michigan',
 'Minnesota',
 'Mississippi',
 'Missouri',
 'Montana',
 'Nebraska',
 'Nevada',
 'New Hampshire',
 'New Mexico',
 'New York',
 'North Carolina',
 'North Dakota',
 'Ohio',
 'Oklahoma',
 'Oregon',
 'Pennsylvania',
 'South Carolina',
 'South Dakota',
 'Tennessee',
 'Texas',
 'Utah',
 'Vermont',
 'Virginia',
 'Washington',
 'West Virginia',
 'Wisconsin',
 'Wyoming']

In [20]:
metroisnotnull= list(data[data.MetroAreaCode.isnull() == False].groupby('State')['State'].count().index)
allstates=list(pd.unique(data.State))
len(allstates)

51

In [21]:
state_whereallareurban=set(allstates)-set(metroisnull)
state_whereallareurban

{'District of Columbia', 'New Jersey', 'Rhode Island'}

In [22]:
states_where_all_are_rural=set(allstates)-set(metroisnotnull)
states_where_all_are_rural

{'Alaska', 'Wyoming'}

#### Evaluating Missing Values
Which region of the United States has the largest proportion of interviewees living in a non-metropolitan area?

In [23]:
rur= (data[data.MetroAreaCode.isnull()]).groupby('Region')['State'].count()
rur

Region
Midwest      10674
Northeast     5609
South         9871
West          8084
Name: State, dtype: int64

In [24]:
all=data.groupby('Region')['State'].count()

In [25]:
rur/all

Region
Midwest      0.347869
Northeast    0.216238
South        0.237844
West         0.243663
Name: State, dtype: float64

#### Perhaps an easier way?? 
* Which STATE has a proportion of interviewees living in a non-metropolitan area closest to 30%?
* Which state has the largest proportion of non-metropolitan interviewees, ignoring states where all interviewees were non-metropolitan?


In [26]:
(data[data.MetroAreaCode.isnull()]).groupby('State').size()/((data[data.MetroAreaCode.isnull()]).groupby('State').size()+(data[data.MetroAreaCode.isnull()==False]).groupby('State').size())

State
Alabama                 0.258721
Alaska                       NaN
Arizona                 0.131545
Arkansas                0.490500
California              0.020484
Colorado                0.129915
Connecticut             0.085684
Delaware                0.233966
District of Columbia         NaN
Florida                 0.039231
Georgia                 0.198432
Hawaii                  0.249166
Idaho                   0.498682
Illinois                0.112219
Indiana                 0.291417
Iowa                    0.486946
Kansas                  0.362274
Kentucky                0.506790
Louisiana               0.161379
Maine                   0.598321
Maryland                0.069375
Massachusetts           0.064922
Michigan                0.178257
Minnesota               0.315068
Mississippi             0.694309
Missouri                0.328671
Montana                 0.836079
Nebraska                0.581324
Nevada                  0.133082
New Hampshire           0.568745
New 

#### Integrating Metropolitan Area Data
When analyzing a variable stored by a numeric code, we will often want to convert it into the values the codes represent. To do this, we will use a dictionary, which maps the the code to the actual value of the variable. We have provided dictionaries MetroAreaCodes.csv and CountryCodes.csv, which respectively map MetroAreaCode and CountryOfBirthCode into their true values. Read these two dictionaries into data frames MetroAreaMap and CountryMap.

To merge in the metropolitan areas, we want to connect the field MetroAreaCode from the CPS data frame with the field Code in MetroAreaMap

In [27]:
metroAC=pd.read_csv('MetroAreaCodes.csv')
cc=pd.read_csv('CountryCodes.csv')
metroAC.shape

(271, 2)

In [28]:
cc.shape

(149, 2)

In [29]:
metroAC.head()

Code                            MetroArea
0   460          Appleton-Oshkosh-Neenah, WI
1  3000    Grand Rapids-Muskegon-Holland, MI
2  3160  Greenville-Spartanburg-Anderson, SC
3  3610                        Jamestown, NY
4  3720           Kalamazoo-Battle Creek, MI

In [30]:
cc.head()

Code                   Country
0    57             United States
1    66                      Guam
2    73               Puerto Rico
3    78      U. S. Virgin Islands
4    96  Other U. S. Island Areas

In [31]:
data.head(2)

PeopleInHousehold Region    State  MetroAreaCode  Age        Married  \
0                  1  South  Alabama        26620.0   85        Widowed   
1                  3  South  Alabama        13820.0   21  Never Married   

      Sex         Education   Race  Hispanic  CountryOfBirthCode  \
0  Female  Associate degree  White         0                  57   
1    Male       High school  Black         0                  57   

       Citizenship EmploymentStatus                            Industry  
0  Citizen, Native          Retired                                 NaN  
1  Citizen, Native       Unemployed  Professional and business services

In [32]:
##merge the data frames:
tmp=data.merge(metroAC,how='left',left_on='MetroAreaCode',right_on='Code',sort=False )
tmp.head(3)

PeopleInHousehold Region    State  MetroAreaCode  Age        Married  \
0                  1  South  Alabama        26620.0   85        Widowed   
1                  3  South  Alabama        13820.0   21  Never Married   
2                  3  South  Alabama        13820.0   37  Never Married   

      Sex         Education   Race  Hispanic  CountryOfBirthCode  \
0  Female  Associate degree  White         0                  57   
1    Male       High school  Black         0                  57   
2  Female       High school  Black         0                  57   

       Citizenship EmploymentStatus                            Industry  \
0  Citizen, Native          Retired                                 NaN   
1  Citizen, Native       Unemployed  Professional and business services   
2  Citizen, Native         Disabled                                 NaN   

      Code              MetroArea  
0  26620.0         Huntsville, AL  
1  13820.0  Birmingham-Hoover, AL  
2  13820.0  Birmingham-Hoover, AL

In [33]:
tmp.shape

(131302, 16)

In [34]:
data.shape

(131302, 14)

#### How many interviewees have a missing value for the new metropolitan area variable? Note that all of these interviewees would have been removed from the merged data frame if we did not include the all.x=TRUE parameter.

In [35]:
tmp.MetroArea.isnull().sum()

34238

In [36]:
#tmp.groupby('MetroArea').size().order()
tmp.groupby('MetroArea').size().sort_values()

MetroArea
Bowling Green, KY                                       29
Ocean City, NJ                                          30
Springfield, OH                                         34
Bloomington-Normal IL                                   40
Valdosta, GA                                            42
Warner Robins, GA                                       42
Tallahassee, FL                                         43
Columbia, MO                                            47
Punta Gorda, FL                                         48
Midland, TX                                             51
Niles-Benton Harbor, MI                                 51
Johnson City, TN                                        52
Santa Fe, NM                                            52
Prescott, AZ                                            54
Vineland-Millville-Bridgeton, NJ                        54
Hickory-Morgantown-Lenoir, NC                           57
Madera, CA                                    

#### Which metropolitan area has the highest proportion of interviewees of Hispanic ethnicity?

In [37]:
tmp['Hispanic'].head(4)

0    0
1    0
2    0
3    0
Name: Hispanic, dtype: int64

In [38]:
(tmp.groupby('MetroArea')['Hispanic'].sum()/tmp.groupby('MetroArea')['Hispanic'].count()).sort_values(ascending=False).head(5)

MetroArea
Laredo, TX                    0.966292
McAllen-Edinburg-Pharr, TX    0.948718
Brownsville-Harlingen, TX     0.797468
El Paso, TX                   0.790984
El Centro, CA                 0.686869
Name: Hispanic, dtype: float64

#### Determine the number of metropolitan areas in the United States from which at least 20% of interviewees are Asian.

In [51]:
#This is really cool:Though there are many mertroareas with NO asians- we can divide by all metro areas and 
#just get Nans for the ones where there are no asians!
(tmp[tmp.Race=='Asian'].groupby('MetroArea').size()/tmp.groupby('MetroArea').size()).sort_values(ascending=False)

MetroArea
Honolulu, HI                                          0.501904
San Francisco-Oakland-Fremont, CA                     0.246753
San Jose-Sunnyvale-Santa Clara, CA                    0.241791
Vallejo-Fairfield, CA                                 0.203008
Fresno, CA                                            0.184818
Warner Robins, GA                                     0.166667
Stockton, CA                                          0.155440
Atlantic City, NJ                                     0.144144
Sacramento-Arden-Arcade-Roseville, CA                 0.142429
San Diego-Carlsbad-San Marcos, CA                     0.142227
Los Angeles-Long Beach-Santa Ana, CA                  0.135056
Olympia, WA                                           0.131313
Salinas, CA                                           0.125000
New York-Northern New Jersey-Long Island, NY-NJ-PA    0.104271
Seattle-Tacoma-Bellevue, WA                           0.099602
Visalia-Porterville, CA                      

#### Determine which metropolitan area has the smallest proportion of interviewees who have received no high school diploma.

In [40]:
tmp.head()


PeopleInHousehold Region    State  MetroAreaCode  Age        Married  \
0                  1  South  Alabama        26620.0   85        Widowed   
1                  3  South  Alabama        13820.0   21  Never Married   
2                  3  South  Alabama        13820.0   37  Never Married   
3                  3  South  Alabama        13820.0   18  Never Married   
4                  3  South  Alabama        26620.0   52        Widowed   

      Sex               Education   Race  Hispanic  CountryOfBirthCode  \
0  Female        Associate degree  White         0                  57   
1    Male             High school  Black         0                  57   
2  Female             High school  Black         0                  57   
3    Male  No high school diploma  Black         0                  57   
4  Female        Associate degree  White         0                  57   

       Citizenship    EmploymentStatus                            Industry  \
0  Citizen, Native             Retired                                 NaN   
1  Citizen, Native          Unemployed  Professional and business services   
2  Citizen, Native            Disabled                                 NaN   
3  Citizen, Native  Not in Labor Force                                 NaN   
4  Citizen, Native            Employed  Professional and business services   

      Code              MetroArea  
0  26620.0         Huntsville, AL  
1  13820.0  Birmingham-Hoover, AL  
2  13820.0  Birmingham-Hoover, AL  
3  13820.0  Birmingham-Hoover, AL  
4  26620.0         Huntsville, AL

In [41]:
tmp.Education.value_counts()


High school                30906
Bachelor's degree          19443
Some college, no degree    18863
No high school diploma     16095
Associate degree            9913
Master's degree             7816
Doctorate degree            1516
Professional degree         1412
Name: Education, dtype: int64

In [42]:
#NA groups in GroupBy are automatically excluded. 
((tmp[tmp.Education=='No high school diploma']).groupby('MetroArea').size()/tmp.groupby('MetroArea').size()).sort_values()



MetroArea
Iowa City, IA                           0.022901
Bowling Green, KY                       0.034483
Kalamazoo-Portage, MI                   0.039370
Champaign-Urbana, IL                    0.040984
Bremerton-Silverdale, WA                0.045977
Jacksonville, NC                        0.047619
Bloomington-Normal IL                   0.050000
Lawrence, KS                            0.051020
Salisbury, MD                           0.054054
Eau Claire, WI                          0.054545
Gainesville, FL                         0.057143
Fort Collins-Loveland, CO               0.058252
Palm Bay-Melbourne-Titusville, FL       0.059524
Madison, WI                             0.059859
Fargo, ND-MN                            0.060185
Altoona, PA                             0.060976
Billings, MT                            0.065327
Ocean City, NJ                          0.066667
Lakeland-Winter Haven, FL               0.067114
Albany-Schenectady-Troy, NY             0.067164
Coeur d'Al

#### Integrating Country of Birth Data
Need to merge in the country of birth information from the CountryMap data frame, 

In [43]:
final=tmp.merge(cc,left_on='CountryOfBirthCode',right_on='Code',sort=False,how='left')

In [44]:
final.drop(['Code_x','Code_y'],axis=1)


PeopleInHousehold Region    State  MetroAreaCode  Age        Married  \
0                       1  South  Alabama        26620.0   85        Widowed   
1                       3  South  Alabama        13820.0   21  Never Married   
2                       3  South  Alabama        13820.0   37  Never Married   
3                       3  South  Alabama        13820.0   18  Never Married   
4                       3  South  Alabama        26620.0   52        Widowed   
5                       3  South  Alabama        26620.0   24  Never Married   
6                       3  South  Alabama        26620.0   26  Never Married   
7                       2  South  Alabama        33660.0   71       Divorced   
8                       2  South  Alabama        33660.0   43       Divorced   
9                       2  South  Alabama        26620.0   52  Never Married   
10                      2  South  Alabama        26620.0   29  Never Married   
11                      1  South  Alabama        33860.0   68        Widowed   
12                      4  South  Alabama        26620.0   24       Divorced   
13                      4  South  Alabama        26620.0    2            NaN   
14                      4  South  Alabama        26620.0    4            NaN   
15                      4  South  Alabama        26620.0   22  Never Married   
16                      2  South  Alabama        13820.0   40  Never Married   
17                      2  South  Alabama        13820.0   13            NaN   
18                      1  South  Alabama        33860.0   42  Never Married   
19                      2  South  Alabama        33660.0   45       Divorced   
20                      2  South  Alabama        33660.0   48       Divorced   
21                      4  South  Alabama        13820.0   78        Widowed   
22                      4  South  Alabama        13820.0   52  Never Married   
23                      4  South  Alabama        13820.0   21  Never Married   
24                      4  South  Alabama        13820.0   35  Never Married   
25                      3  South  Alabama        33860.0   30        Married   
26                      3  South  Alabama        33860.0   33        Married   
27                      3  South  Alabama        33860.0    2            NaN   
28                      3  South  Alabama        13820.0   60        Widowed   
29                      3  South  Alabama        13820.0   24  Never Married   
...                   ...    ...      ...            ...  ...            ...   
131272                  1   West  Wyoming            NaN   43  Never Married   
131273                  1   West  Wyoming            NaN   32      Separated   
131274                  4   West  Wyoming            NaN   34  Never Married   
131275                  4   West  Wyoming            NaN    9            NaN   
131276                  4   West  Wyoming            NaN   11            NaN   
131277                  4   West  Wyoming            NaN   32       Divorced   
131278                  5   West  Wyoming            NaN   46        Married   
131279                  5   West  Wyoming            NaN   63        Married   
131280                  5   West  Wyoming            NaN   13            NaN   
131281                  5   West  Wyoming            NaN    4            NaN   
131282                  5   West  Wyoming            NaN    9            NaN   
131283                  2   West  Wyoming            NaN   19        Married   
131284                  2   West  Wyoming            NaN   21        Married   
131285                  1   West  Wyoming            NaN   68       Divorced   
131286                  2   West  Wyoming            NaN   60        Married   
131287                  2   West  Wyoming            NaN   61        Married   
131288                  2   West  Wyoming            NaN   80        Married   
131289                  2   West  Wyoming            NaN   80        Married   
131290  

#### How many interviewees have a missing value for the new country of birth variable?

In [45]:
final['Country'].isnull().sum()

176

#### Among all interviewees born outside of North America, which country was the most common place of birth?

In [46]:
(final[final.Country != 'United States'])['Country'].value_counts().head()

Mexico         3921
Philippines     839
India           770
China           581
Puerto Rico     518
Name: Country, dtype: int64

#### What proportion of the interviewees from the "New York-Northern New Jersey-Long Island, NY-NJ-PA" metropolitan area have a country of birth that is not the United States? 

In [47]:
float(final[final.MetroArea=='New York-Northern New Jersey-Long Island, NY-NJ-PA'][final.Country != 'United States']['MetroAreaCode'].count())/final[final.MetroArea=='New York-Northern New Jersey-Long Island, NY-NJ-PA']['MetroAreaCode'].count()

/Users/priyadesai/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  if __name__ == '__main__':


0.30929931595489

#### Which metropolitan area has the largest number (note -- not proportion) of interviewees with a country of birth: 
* in India?
* in Brazil?
* in Somalia?

In [48]:
(final[final.Country=='India']).groupby('MetroArea').size().sort_values(ascending=False).head()

MetroArea
New York-Northern New Jersey-Long Island, NY-NJ-PA    96
Washington-Arlington-Alexandria, DC-VA-MD-WV          50
Philadelphia-Camden-Wilmington, PA-NJ-DE              32
Chicago-Naperville-Joliet, IN-IN-WI                   31
Detroit-Warren-Livonia, MI                            30
dtype: int64

In [49]:
final[final.Country=='Brazil'].groupby('MetroArea').size().sort_values(ascending=False).head()

MetroArea
Boston-Cambridge-Quincy, MA-NH                        18
Miami-Fort Lauderdale-Miami Beach, FL                 16
Los Angeles-Long Beach-Santa Ana, CA                   9
Washington-Arlington-Alexandria, DC-VA-MD-WV           8
New York-Northern New Jersey-Long Island, NY-NJ-PA     7
dtype: int64

In [50]:
final[final.Country=='Somalia'].groupby('MetroArea').size().sort_values(ascending=False).head()

MetroArea
Minneapolis-St Paul-Bloomington, MN-WI    17
St. Cloud, MN                              7
Seattle-Tacoma-Bellevue, WA                7
Phoenix-Mesa-Scottsdale, AZ                7
Fargo, ND-MN                               5
dtype: int64